websockets.md

```diff
@@ -0,0 +1,101 @@
+
+In case of failure, HTTP status code can help to distinguish between retriable and non-retriable errors (e.g. server doesn't support WebSockets at all, or it was just a small transient error). Headers might also contain additional information on how to handle the situation. The headers are also useful even in case of a success WebSocket connect, e.g. they can contain a token tied to a session, or some info related to subprotocol version, or that the server can go down soon, etc.
+
+.NET 7 adds a setting [ClientWebSocketOptions.CollectHttpResponseDetails](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocketoptions.collecthttpresponsedetails?view=net-7.0#system-net-websockets-clientwebsocketoptions-collecthttpresponsedetails) that enables collecting upgrade response details in `ClientWebSocket` instance during `ClientWebSocket.ConnectAsync` call. You can later access the data using [ClientWebSocket.HttpStatusCode](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.httpstatuscode?view=net-7.0) and [ClientWebSocket.HttpResponseHeaders](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.httpresponseheaders?view=net-7.0) properties, even in case of `ClientWebSocket.ConnectAsync` throwing an exception. Note that in the exception case, the information might be unavailable, e.g. if the server never responded to the request.
+
+Also note that in case of a successful connect and after consuming `HttpResponseHeaders` data, you can reduce `ClientWebSocket`'s memory footprint by setting `ClientWebSocket.HttpResponseHeaders` property to `null`.
+
+```c#
+var ws = new ClientWebSocket();
+ws.Options.CollectHttpResponseDetails = true;
+try
+{
+    await ws.ConnectAsync(uri, cancellationToken);
+    // success scenario
+    ProcessSuccess(ws.HttpResponseHeaders);
+    ws.HttpResponseHeaders = null; // clean up (if needed)
+}
+catch (WebSocketException)
+{
+    // failure scenario
+    if (ws.HttpStatusCode != 0)
+    {
+        ProcessFailure(ws.HttpStatusCode, ws.HttpResponseHeaders);
+    }
+}
+```
+
+## Providing external HTTP client
+
+In a default case, `ClientWebSocket` would use a cached static `HttpMessageInvoker` instance to execute HTTP Upgrade requests. However, if options such as `ClientWebSocketOptions.Proxy`, `ClientWebSocketOptions.ClientCertificates` or `ClientWebSocketOptions.Cookies` are passed, `HttpMessageInvoker`s with such parameters are not safe to be reused, so they would be created each time per `ClientWebSocket.ConnectAsync` call. This results in more allocations that could be necessary and makes potential reuse of `HttpMessageInvoker` instances and established HTTP connections impossible.
```

Comment:
```suggestion
In the default case, `ClientWebSocket` uses a cached static `HttpMessageInvoker` instance to execute HTTP Upgrade requests. However, there are `ClientWebSocket` [options](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.options?view=net-7.0#system-net-websockets-clientwebsocket-options) that prevent caching the invoker, such as `ClientWebSocketOptions.Proxy`, `ClientWebSocketOptions.ClientCertificates` or `ClientWebSocketOptions.Cookies`. `HttpMessageInvoker` instance with these parameters is not safe to be reused and needs be created each time `ClientWebSocket.ConnectAsync` is called. This results in many unnecessary allocations and makes reuse of `HttpMessageInvoker` connection pool impossible.
```

Reply:
Do I get it right that the invoker/client is only used to do Upgrade request? What the heck is Upgrade request? GET with Upgrade header?
Would it be better to introduce it as "initiating HTTP request, aka Upgrade request", or something in this regard.

Reply:
You're right, and what is worse, RFC uses term "Opening Handshake", so "Upgrade request" is just what people usually call it 😄 though "Opening Handshake" sounds even more unclear. I'll rephrase that somehow